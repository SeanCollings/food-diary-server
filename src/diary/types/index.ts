import { MAX_CALENDAR_ENTRIES_RANGE } from '@/lib/constants/validation/validation.constants';
import { Type } from 'class-transformer';
import { IsDateString, IsNotEmpty, IsNumber, Max } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export class GetDiaryEntriesQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class GetCalendarEntriesQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;

  @IsNotEmpty()
  @IsNumber()
  @Max(MAX_CALENDAR_ENTRIES_RANGE)
  @Type(() => Number)
  months: string;
}

export type MealColumnKeys =
  | 'mealBreakfast'
  | 'mealSnack1'
  | 'mealLunch'
  | 'mealSnack2'
  | 'mealDinner';

export type HasMealColumnKeys =
  | 'hasMealBreakfast'
  | 'hasMealSnack1'
  | 'hasMealLunch'
  | 'hasMealSnack2'
  | 'hasMealDinner';
