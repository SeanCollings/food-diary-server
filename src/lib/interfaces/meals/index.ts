export enum EMealType {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type TMealType = `${EMealType}`;

export type TSelectedEmoji = {
  name: string;
  nativeSkin: string;
} | null;

export interface IMealContent {
  id: string;
  emoji?: TSelectedEmoji;
  serving?: string;
  measurement?: string;
  food: string;
  description?: string;
}

export interface IMealTypeContent {
  contents: IMealContent[];
}

export type TMealCard = {
  [key in TMealType]?: IMealTypeContent;
};
