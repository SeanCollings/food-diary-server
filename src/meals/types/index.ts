import { IsDateString, IsNotEmpty } from 'class-validator';
import { Request } from 'express';

// https://stackoverflow.com/questions/50160662/convert-nested-array-type

export class CreateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class UpdateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class DeleteMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export enum MealTypesEnum {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type MealTypes = `${MealTypesEnum}`;

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export interface MealContent {
  id: string;
  food: string;
  serving?: string;
  quantity?: string;
  description?: string;
  emoji?: { name: string; nativeSkin: string } | null;
}

export type SetFieldOld<V, O extends object> = {
  [K in keyof O]: O[K] extends object ? SetFieldOld<V, O[K]> : V;
};

type Primitive = string | number | boolean | null | undefined;

type SetField<V, O> = O extends Primitive
  ? V
  : O extends (infer U)[]
  ? SetField<V, U>
  : { [K in keyof O]: SetField<V, O[K]> };

type RequireOne<T> = T & { [P in keyof T]: Required<Pick<T, P>> }[keyof T];
