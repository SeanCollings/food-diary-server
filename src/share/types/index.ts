import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import {
  IsDateString,
  IsNotEmpty,
  IsString,
  IsUUID,
  MaxLength,
  MinLength,
} from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetSharedSummaryQuery {
  @IsNotEmpty()
  @MinLength(36)
  @MaxLength(36)
  @IsString()
  @IsUUID()
  link: string;

  @IsNotEmpty()
  @IsDateString()
  dateFrom: string;

  @IsNotEmpty()
  @IsDateString()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
