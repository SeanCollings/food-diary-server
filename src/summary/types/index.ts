import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import { IsDateString, IsNotEmpty } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetUserSummaryQuery {
  @IsDateString()
  @IsNotEmpty()
  dateFrom: string;

  @IsDateString()
  @IsNotEmpty()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
