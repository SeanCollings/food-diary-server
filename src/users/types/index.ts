import { User } from '@prisma/client';

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export type PreferenceColumnNames = keyof Pick<
  User,
  | 'userPreferenceShowDayStreak'
  | 'userPreferenceShowWeeklyExcercise'
  | 'userPreferenceShowWeeklyWater'
>;

export type UserWithShareLink =
  | (User & {
      shareLink: {
        link: string;
        isShared: boolean;
      } | null;
    })
  | null;
