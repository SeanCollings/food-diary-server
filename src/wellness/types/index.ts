export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export enum WellnessTypesEnum {
  WATER = 'water',
  TEA_COFFEE = 'tea_coffee',
  ALCOHOL = 'alcohol',
  EXCERCISE = 'excercise',
}
export type WellnessTypes = `${WellnessTypesEnum}`;
