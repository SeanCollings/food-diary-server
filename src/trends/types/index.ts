import { IsIn, IsNotEmpty, IsString } from 'class-validator';
import { WellnessTypesEnum } from '@/wellness/types';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export type TrendType = 'week' | 'month';

export class GetTrendsQuery {
  @IsString()
  @IsNotEmpty()
  @IsIn(['week', 'month'])
  type: TrendType;
}

export type BeverageTypes = Exclude<
  WellnessTypesEnum,
  WellnessTypesEnum.EXCERCISE
>;
