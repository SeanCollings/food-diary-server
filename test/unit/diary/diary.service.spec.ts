import { DiaryService } from '@/diary/diary.service';
import { MealContent } from '@/meals/types';
import { PrismaService } from '@/prisma.service';
import { Test, TestingModule } from '@nestjs/testing';
import { DiaryDay, PrismaClient } from '@prisma/client';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';

describe('DiaryService', () => {
  const mockUserId = 'mock_user_id';
  const mockDate = '2023-04-28';

  let service: DiaryService;
  let prisma: DeepMockProxy<PrismaClient>;

  const mockDiaryDays: Partial<DiaryDay>[] = [
    {
      id: '1',
      userId: mockUserId,
      date: mockDate,
      hasMealBreakfast: true,
      hasMealSnack1: true,
      hasMealDinner: true,
      wellnessWater: 2,
      wellnessAlcohol: 1,
      wellnessExcercise: '01:15',
      wellnessExcerciseDetails: 'mock exercise details',
      mealBreakfast: [
        { id: '1', food: 'mock_food_1.1', description: 'mock_description_1.1' },
      ] as MealContent[] as any,
      mealSnack1: [
        { id: '2.1', food: 'mock_food_2.1' },
      ] as MealContent[] as any,
      mealDinner: [
        { id: '3.1', food: 'mock_food_3.1' },
      ] as MealContent[] as any,
    },
    {
      id: '2',
      userId: mockUserId,
      date: '2023-04-27',
      hasMealBreakfast: true,
      hasMealLunch: true,
      wellnessWater: 4,
      wellnessTeaCoffee: 2,
      mealBreakfast: [
        { id: '1.2', food: 'mock_food_1.2', serving: 'mock_serving_1.2' },
      ] as MealContent[] as any,
      mealDinner: [
        { id: '2.2', food: 'mock_food_2.2' },
      ] as MealContent[] as any,
    },
  ];

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [DiaryService, PrismaService],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .compile();

    service = module.get(DiaryService);
    prisma = module.get(PrismaService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getDiaryEntries', () => {
    it('should get diary entries for a specified date', async () => {
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDays as any);

      const result = await service.getDiaryEntries(mockUserId, mockDate);

      expect(result).toMatchSnapshot();
    });

    it('should cater for no diary-day for specified date', async () => {
      prisma.diaryDay.findMany.mockResolvedValue([]);

      const result = await service.getDiaryEntries(mockUserId, mockDate);

      expect(result).toMatchSnapshot();
    });
  });

  describe('getCalendarEntries', () => {
    it('should get calendar entries for a date and number of months back', async () => {
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDays as any);

      const result = await service.getCalendarEntries(
        mockUserId,
        mockDate,
        '3',
      );

      expect(result).toMatchSnapshot();
    });

    it('should cater for no diary-days returned', async () => {
      prisma.diaryDay.findMany.mockResolvedValue([]);

      const result = await service.getCalendarEntries(
        mockUserId,
        mockDate,
        '3',
      );

      expect(result).toMatchSnapshot();
    });
  });
});
