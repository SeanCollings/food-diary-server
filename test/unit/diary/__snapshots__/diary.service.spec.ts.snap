// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiaryService getCalendarEntries should cater for no diary-days returned 1`] = `
{
  "entries": {},
  "monthRange": {
    "1-2023": true,
    "2-2023": true,
    "3-2023": true,
  },
}
`;

exports[`DiaryService getCalendarEntries should get calendar entries for a date and number of months back 1`] = `
{
  "entries": {
    "3-2023": [
      28,
      27,
    ],
  },
  "monthRange": {
    "1-2023": true,
    "2-2023": true,
    "3-2023": true,
  },
}
`;

exports[`DiaryService getDiaryEntries should cater for no diary-day for specified date 1`] = `null`;

exports[`DiaryService getDiaryEntries should get diary entries for a specified date 1`] = `
{
  "meals": {
    "2023-04-27": {
      "breakfast": {
        "contents": [
          {
            "food": "mock_food_1.2",
            "id": "1.2",
            "quantity": "mock_quantity_1.2",
          },
        ],
      },
      "dinner": {
        "contents": [
          {
            "food": "mock_food_2.2",
            "id": "2.2",
          },
        ],
      },
      "lunch": {
        "contents": [],
      },
      "snack_1": {
        "contents": [],
      },
      "snack_2": {
        "contents": [],
      },
    },
    "2023-04-28": {
      "breakfast": {
        "contents": [
          {
            "description": "mock_description_1.1",
            "food": "mock_food_1.1",
            "id": "1",
          },
        ],
      },
      "dinner": {
        "contents": [
          {
            "food": "mock_food_3.1",
            "id": "3.1",
          },
        ],
      },
      "lunch": {
        "contents": [],
      },
      "snack_1": {
        "contents": [
          {
            "food": "mock_food_2.1",
            "id": "2.1",
          },
        ],
      },
      "snack_2": {
        "contents": [],
      },
    },
  },
  "wellness": {
    "2023-04-27": {
      "alcohol": {
        "value": 0,
      },
      "excercise": {
        "details": "",
        "time": "00:00",
      },
      "tea_coffee": {
        "value": 2,
      },
      "water": {
        "value": 4,
      },
    },
    "2023-04-28": {
      "alcohol": {
        "value": 1,
      },
      "excercise": {
        "details": "mock exercise details",
        "time": "01:15",
      },
      "tea_coffee": {
        "value": 0,
      },
      "water": {
        "value": 2,
      },
    },
  },
}
`;
